File: DFRobot_STCC4.cpp
/*!
 * @file DFRobot_STCC4.cpp
 * @brief Implement the basic structure of class DFRobot_STCC4
 * @n It is possible to measure the concentration of carbon dioxide in the air.
 * @n If the humidity and temperature sensor is connected, temperature and humidity can also be obtained.
 * @n Only supports IIC communication interface
 * @copyright	Copyright (c) 2025 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @license The MIT License (MIT)
 * @author [lbx](liubx8023@gmail.com)
 * @version V1.0
 * @date 2025-08-15
 * @url https://github.com/DFRobot/DFRobot_STCC4
 */

#include <DFRobot_STCC4.h>
DFRobot_STCC4::DFRobot_STCC4(void) {}

DFRobot_STCC4::~DFRobot_STCC4(void) {}

uint8_t DFRobot_STCC4::calculationCRC(uint16_t *data, size_t length)
{
  uint8_t crc = 0xFF;

  for (size_t i = 0; i < length; i++) {
    uint8_t high_byte = (data[i] >> 8) & 0xFF; 
    uint8_t low_byte = data[i] & 0xFF;          
    crc ^= high_byte;
    for (uint8_t bit = 0; bit < 8; bit++) {
        if (crc & 0x80) {
            crc = (crc << 1) ^ 0x31;
        } else {
            crc <<= 1;
        }
    }
    crc ^= low_byte;
    for (uint8_t bit = 0; bit < 8; bit++) {
        if (crc & 0x80) {
            crc = (crc << 1) ^ 0x31;
        } else {
            crc <<= 1;
        }
    }
  }

  return crc;
}

uint32_t DFRobot_STCC4::getID(void)  
{
  for (uint8_t i = 0; i < 5; i++){
    delay(200);
    uint8_t rBuf[18] = {0};
    uint32_t id;
    writeCMD16(STCC4_GET_ID);
    readData(rBuf, 18);
    
    uint16_t data1 = ((uint16_t)rBuf[0] << 8) | rBuf[1];
    uint16_t data2 = ((uint16_t)rBuf[3] << 8) | rBuf[4];
    uint8_t crc1 = rBuf[2];
    uint8_t crc2 = rBuf[5];
    
    uint8_t calculatedCrc1 = calculationCRC(&data1, 1);
    uint8_t calculatedCrc2 = calculationCRC(&data2, 1);
    
    if(crc1 == calculatedCrc1 || crc2 == calculatedCrc2) {
      id = ((uint32_t)rBuf[0] << 24) | ((uint32_t)rBuf[1] << 16) | ((uint32_t)rBuf[3] << 8) | (uint32_t)rBuf[4];
      return id;
    }
  }
  
  return 0;
}

bool DFRobot_STCC4::startMeasurement(void)
{
  if(writeCMD16(STCC4_START_CONT_MEASURE)){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::stopMeasurement(void)
{
  if(writeCMD16(STCC4_STOP_CONT_MEASURE)){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::measurement(uint16_t* co2Concentration, 
                                        float* temperature, 
                                        float* humidity, 
                                        uint16_t* sensorStatus)
{
  uint8_t rBuf[12];
  if(writeCMD16(STCC4_READ_MEASURE) != ERR_OK){
    return false;
  }
  if(readData(rBuf, 12) != 12){
    return false;
  }

  *co2Concentration = (rBuf[0] << 8) | rBuf[1];
  int16_t tempRaw = (rBuf[3] << 8) | rBuf[4];
  *temperature = -45.0 + ((175.0 * tempRaw) / 65535.0);
  uint16_t humRaw = (rBuf[6] << 8) | rBuf[7];
  *humidity = -6.0 + ((125.0 * humRaw) / 65535.0);
  *sensorStatus = (rBuf[9] << 8) | rBuf[10];

  return true;
}

bool DFRobot_STCC4::setRHTcompensation(uint16_t temperature, uint16_t humidity)
{
  if(temperature < 10 || temperature > 40 || humidity < 20 || humidity > 80)
  {
    return false;
  }
  uint16_t wBuf[2];
  temperature = (temperature + 45) * 65535 / 175; 
  humidity = (humidity + 6) * 65535 / 125; 
  wBuf[0] = temperature;
  wBuf[1] = humidity;
  if(writeData(STCC4_SET_RHT_COMPENSATION, wBuf, 2) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::setPressureCompensation(uint16_t pressure)
{
  if(pressure < 400 || pressure > 1100)
  {
    return false;
  }
  pressure = pressure  * 50;
  if(writeData(STCC4_SET_PRESSURE_COMPENSATION, &pressure, 1) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::singleMeasurement(void)
{
  if(writeCMD16(STCC4_SINGLE_SHOT) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::sleep(void)
{
  if(writeCMD16(STCC4_SLEEP) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::wakeup(void)
{
  if(writeCMD8(STCC4_WAKEUP) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::softRest(void)
{
  if(writeCMD8(STCC4_SOFT_RESET) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::factoryReset(void)
{
  uint8_t rBuf[2];
  if(writeCMD16(STCC4_FACTORY_RESET) != ERR_OK){
    return false;
  }
  if(readData(rBuf, 2) != 2)
    return false;
  
  uint16_t response = (rBuf[0] << 8) | rBuf[1];
  if(response == 0)
    return true;
  
  return false;
}

bool DFRobot_STCC4::enableTestingMode(void)
{
  if(writeCMD16(STCC4_ENABLE_TESTING_MODE) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::disableTestingMode(void)
{
  if(writeCMD16(STCC4_DISABLE_TESTING_MODE) != ERR_OK){
    return false;
  }

  return true;
}

bool DFRobot_STCC4::forcedRecalibration(uint16_t targetPpm, uint16_t* frcCorrection)
{
  if(targetPpm > 32000)
  {
    return false;
  }
  uint16_t wBuf[1];
  uint8_t rBuf[3];
  wBuf[0] = targetPpm;
  if(writeData(STCC4_FORC_CALIBRATION, wBuf, 1) != ERR_OK){
    return false;
  }
  delay(200);
  if(readData(rBuf, 3) != 3){
    return false;
  }
  *frcCorrection = rBuf[0] << 8 | rBuf[1];

  return true;
}

bool DFRobot_STCC4_I2C::begin(void) 
{
  _pWire->begin();
  _pWire->beginTransmission(_deviceAddr);
  
  if(_pWire == NULL)
  {
    Serial.println("_pWire == NULL");
    return ERR_DATA_BUS;
  }
  return _pWire->endTransmission() == 0 ? true : false; 
}

DFRobot_STCC4_I2C::DFRobot_STCC4_I2C(TwoWire *pWire, uint8_t addr) {
  _pWire = pWire;
  _deviceAddr = addr;
  _pWire->setClock(100000);
}

bool DFRobot_STCC4_I2C::writeData(uint16_t cmd, uint16_t * pBuf, size_t size)
{
  uint8_t buf[2];
  buf[0] = (cmd >> 8) & 0xFF;
  buf[1] = cmd & 0xFF;

  if(_pWire == NULL) {
    Serial.println("_pWire == NULL");
  }

  _pWire->beginTransmission(_deviceAddr);
  _pWire->write(buf[0]);
  _pWire->write(buf[1]);

  for (size_t i = 0; i < size; i++) {
    buf[0] = (pBuf[i] >> 8) & 0xFF;
    buf[1] = pBuf[i] & 0xFF;
    _pWire->write(buf[0]);
    _pWire->write(buf[1]);
    _pWire->write(calculationCRC(&pBuf[i], 1));
  }

  if(_pWire->endTransmission() != 0) {
    //DBG("I2C write error");
    return ERR_DATA_WRITE;
  }

  return ERR_OK;
}

bool DFRobot_STCC4_I2C::writeCMD8(uint8_t cmd)
{
  _pWire->beginTransmission(_deviceAddr);
  _pWire->write(cmd);

  if(_pWire->endTransmission() != 0) {
    //DBG("I2C write error");
    return ERR_DATA_WRITE;
  }

  return ERR_OK;
}

bool DFRobot_STCC4_I2C::writeCMD16(uint16_t cmd)
{
  uint8_t buf[2];
  buf[0] = (cmd >> 8) & 0xFF;
  buf[1] = cmd & 0xFF;

  if(_pWire == NULL) {
    Serial.println("_pWire == NULL");
  }

  _pWire->beginTransmission(_deviceAddr);
  _pWire->write(buf[0]);
  _pWire->write(buf[1]);

  if(_pWire->endTransmission() != 0) {
    //DBG("I2C write error");
    return ERR_DATA_WRITE;
  }

  return ERR_OK;
}

size_t DFRobot_STCC4_I2C::readData(uint8_t * pBuf, size_t size)
{
  size_t ret = 0;

  if(_pWire == NULL) {
    Serial.println("_pWire == NULL");
  }

  ret = _pWire->requestFrom(_deviceAddr, (uint8_t) size);
  for (size_t i = 0; i < ret; i++) {
    pBuf[i] = _pWire->read();
  }

  return ret;
}

File: python/raspberrypi/DFRobot_STCC4.py
"""!
    * @file DFRobot_STCC4.py
    * @brief Declaration the basic structure of class DFRobot_STCC4
    * @n It is possible to measure the concentration of carbon dioxide in the air.
    * @n If the humidity and temperature sensor is connected, temperature and humidity can also be obtained.
    * @n Only supports IIC communication interface
    * @copyright	Copyright (c) 2025 DFRobot Co.Ltd (http://www.dfrobot.com)
    * @license The MIT License (MIT)
    * @author [lbx](liubx8023@gmail.com)
    * @version V1.0
    * @date 2025-08-15
    * @url https://github.com/DFRobot/DFRobot_STCC4
 """

import smbus2
import time
from typing import Optional, Tuple, Union
 
class DFRobot_STCC4:
    """Base class for DFRobot STCC4 CO2 sensor"""
    
    ERR_OK = 0
    ERR_DATA_BUS = 1
    ERR_DATA_READ = 2
    ERR_DATA_WRITE = 3
    ERR_IC_VERSION = 4
    
    # Sensor commands
    STCC4_GET_ID = 0x365B
    STCC4_START_CONT_MEASURE = 0x218B
    STCC4_STOP_CONT_MEASURE = 0x3F86
    STCC4_READ_MEASURE = 0xEC05
    STCC4_SET_RHT_COMPENSATION = 0xE000
    STCC4_SET_PRESSURE_COMPENSATION = 0xE016
    STCC4_SINGLE_SHOT = 0x219D
    STCC4_SLEEP = 0x3650
    STCC4_WAKEUP = 0x00
    STCC4_SOFT_RESET = 0x06
    STCC4_FACTORY_RESET = 0x3632
    STCC4_ENABLE_TESTING_MODE = 0x3FBC
    STCC4_DISABLE_TESTING_MODE = 0x3F3D
    STCC4_FORC_CALIBRATION = 0x362F
 
    def __init__(self):
        """Constructor"""
        pass
 
    def calculation_crc(self, data: Union[list, tuple]) -> int:
        """
        Calculate CRC for the data
        :param data: List or tuple of 16-bit integers
        :return: Calculated CRC value
        """
        crc = 0xFF
        
        for value in data:
            high_byte = (value >> 8) & 0xFF
            low_byte = value & 0xFF
            
            # Process high byte
            crc ^= high_byte
            for _ in range(8):
                if crc & 0x80:
                    crc = (crc << 1) ^ 0x31
                else:
                    crc <<= 1
                crc &= 0xFF  # Keep it as 8-bit
            
            # Process low byte
            crc ^= low_byte
            for _ in range(8):
                if crc & 0x80:
                    crc = (crc << 1) ^ 0x31
                else:
                    crc <<= 1
                crc &= 0xFF  # Keep it as 8-bit
        
        return crc
 
    def get_id(self) -> Optional[bytes]:
        """
        Get the sensor ID
        :return: Sensor ID as bytes if successful, None otherwise
        """
        raise NotImplementedError
 
    def start_measurement(self) -> bool:
        """
        Start continuous measurement
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def stop_measurement(self) -> bool:
        """
        Stop continuous measurement
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def measurement(self) -> Optional[Tuple[int, float, float, int]]:
        """
        Read measurement data
        :return: Tuple of (co2_concentration, temperature, humidity, sensor_status).
        co2_concentration : CO2 concentration
        temperature : Temperature
        humidity : Humidity
        sensor_status : Sensor status
        None : error
        """
        raise NotImplementedError
 
    def set_rht_compensation(self, temperature: int, humidity: int) -> bool:
        """
        Set temperature and humidity compensation
        :param temperature: Temperature compensation value, range of 10 to 40 degrees Celsius.
        :param humidity: Humidity compensation value, range of 20 to 80 percent relative humidity.
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def set_pressure_compensation(self, pressure: int) -> bool:
        """
        Set pressure compensation
        :param pressure: Pressure compensation value, range of 400 to 1100 hPa
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def single_measurement(self) -> bool:
        """
        Perform a single shot measurement
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def fall_asleep(self) -> bool:
        """
        Put the sensor into sleep mode
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def wakeup(self) -> bool:
        """
        Wake up the sensor from sleep mode
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def soft_reset(self) -> bool:
        """
        Perform a soft reset of the sensor
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def factory_reset(self) -> bool:
        """
        Perform a factory reset of the sensor
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def enable_testing_mode(self) -> bool:
        """
        Enable testing mode
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def disable_testing_mode(self) -> bool:
        """
        Disable testing mode
        :return: True if successful, False otherwise
        """
        raise NotImplementedError
 
    def forced_recalibration(self, target_ppm: int) -> Optional[int]:
        """
        Perform forced recalibration
        :param target_ppm: Target PPM value for recalibration, must be between 0 and 32000 ppm.
        :return: Correction value if successful, None otherwise
        """
        raise NotImplementedError
 
 
class DFRobot_STCC4_I2C(DFRobot_STCC4):
    """I2C implementation of DFRobot STCC4 CO2 sensor"""
    
    DEFAULT_I2C_ADDR = 0x64
    I2C_BUS = 1  # Raspberry Pi uses bus 1 for I2C
 
    def __init__(self, addr: int = DEFAULT_I2C_ADDR):
        """
        Constructor for I2C implementation
        :param addr: I2C address of the sensor
        """
        super().__init__()
        self._device_addr = addr
        try:
            self._bus = smbus2.SMBus(self.I2C_BUS)
        except Exception as e:
            self._bus = None
 
    def _write_cmd16(self, cmd: int) -> bool:
        """
        Write a 16-bit command to the sensor
        :param cmd: Command to write
        :return: True if successful, False otherwise
        """
        if self._bus is None:
            return False
            
        try:
            # Split 16-bit command into two bytes (big endian)
            bytes_to_send = [(cmd >> 8) & 0xFF, cmd & 0xFF]
            self._bus.write_i2c_block_data(self._device_addr, bytes_to_send[0], [bytes_to_send[1]])
            return True
        except Exception as e:
            return False
 
    def _write_cmd8(self, cmd: int) -> bool:
        """
        Write an 8-bit command to the sensor
        :param cmd: Command to write
        :return: True if successful, False otherwise
        """
        if self._bus is None:
            return False
            
        try:
            self._bus.write_byte(self._device_addr, cmd)
            return True
        except Exception as e:
            return False
 
    def _write_data(self, cmd: int, data: Union[list, tuple]) -> bool:
        """
        Write data to the sensor
        :param cmd: Command to write before data
        :param data: List or tuple of 16-bit integers to write
        :return: True if successful, False otherwise
        """
        if not self._write_cmd16(cmd):
            return False
            
        try:
            for value in data:
                # Split value into two bytes
                high_byte = (value >> 8) & 0xFF
                low_byte = value & 0xFF
                
                # Calculate CRC for this value
                crc = self.calculation_crc([value])
                
                # Write data bytes and CRC

                # self._bus.write_i2c_block_data(
                #     self._device_addr, 
                #     high_byte, 
                #     [low_byte, crc]
                # )
                self._bus.write_byte(self._device_addr, high_byte)
                self._bus.write_byte(self._device_addr, low_byte)
                self._bus.write_byte(self._device_addr, crc)
                # Small delay between writes
                time.sleep(0.01)
            
            return True
        except Exception as e:
            return False
 
    def _read_data(self, cmd: int, length: int) -> Optional[bytes]:
        """
        Read data from the sensor
        :param cmd: Command to write before reading
        :param length: Number of bytes to read
        :return: Read bytes if successful, None otherwise
        """
        if not self._write_cmd16(cmd):
            return None
            
        try:
            # Read the data
            data = self._bus.read_i2c_block_data(self._device_addr, 0, length)
            return bytes(data)
        except Exception as e:
            return None
 
    def get_id(self):
        """
        git sensor id
        Returns:
            int: 32-bit sensor ID
        """
        for i in range(5):
            r_buf = self._read_data(self.STCC4_GET_ID, 18)
            id_value = (r_buf[0] << 24) | (r_buf[1] << 16) | (r_buf[3] << 8) | r_buf[4]
            if id_value == 0x901018A:
                return id_value
            time.sleep(0.2)

        return id_value
 
    def start_measurement(self) -> bool:
        """Start continuous measurement"""
        if not self._write_cmd16(self.STCC4_START_CONT_MEASURE):
            return False
        time.sleep(1)
        return True
 
    def stop_measurement(self) -> bool:
        """Stop continuous measurement"""
        if not self._write_cmd16(self.STCC4_STOP_CONT_MEASURE):
            return False
        time.sleep(1)
        return True
 
    def measurement(self) -> Optional[Tuple[int, float, float, int]]:
        """Read measurement data"""
        raw_data = self._read_data(self.STCC4_READ_MEASURE, 12)
        if raw_data is None or len(raw_data) < 12:
            return None
            
        # Parse CO2 concentration
        co2_concentration = (raw_data[0] << 8) | raw_data[1]
        
        # Parse temperature (raw value to °C)
        temp_raw = (raw_data[3] << 8) | raw_data[4]
        temperature = -45.0 + ((175.0 * temp_raw) / 65535.0)
        
        # Parse humidity (raw value to %RH)
        hum_raw = (raw_data[6] << 8) | raw_data[7]
        humidity = -6.0 + ((125.0 * hum_raw) / 65535.0)
        
        # Parse sensor status
        sensor_status = (raw_data[9] << 8) | raw_data[10]
        
        return (co2_concentration, temperature, humidity, sensor_status)
 
    def set_rht_compensation(self, temperature: float, humidity: float) -> bool:
        """Set temperature and humidity compensation"""
        if temperature < 10 or temperature > 40 or humidity < 20 or humidity > 80:
            return False
        # Convert temperature to raw value
        temp_raw = int((temperature + 45) * 65535 / 175)
        # Convert humidity to raw value
        hum_raw = int((humidity + 6) * 65535 / 125)
        
        return self._write_data(self.STCC4_SET_RHT_COMPENSATION, [temp_raw, hum_raw])
 
    def set_pressure_compensation(self, pressure: int) -> bool:
        """Set pressure compensation"""
        if pressure < 400 or pressure > 1100:
            return False
        pressure_raw = pressure * 50
        return self._write_data(self.STCC4_SET_PRESSURE_COMPENSATION, [pressure_raw])
 
    def single_measurement(self) -> bool:
        """Perform single shot measurement"""
        return self._write_cmd16(self.STCC4_SINGLE_SHOT)

    def fall_asleep(self) -> bool:
        """Put sensor to sleep"""
        return self._write_cmd16(self.STCC4_SLEEP)
 
    def wakeup(self) -> bool:
        """Wake up sensor"""
        return self._write_cmd8(self.STCC4_WAKEUP)
 
    def soft_reset(self) -> bool:
        """Perform soft reset"""
        return self._write_cmd8(self.STCC4_SOFT_RESET)
 
    def factory_reset(self) -> bool:
        """Perform factory reset"""
        if not self._write_cmd16(self.STCC4_FACTORY_RESET):
            return False
            
        raw_data = self._read_data(self.STCC4_FACTORY_RESET, 2)
        if raw_data is None or len(raw_data) < 2:
            return False
            
        response = (raw_data[0] << 8) | raw_data[1]
        return response == 0
 
    def enable_testing_mode(self) -> bool:
        """Enable testing mode"""
        return self._write_cmd16(self.STCC4_ENABLE_TESTING_MODE)
 
    def disable_testing_mode(self) -> bool:
        """Disable testing mode"""
        return self._write_cmd16(self.STCC4_DISABLE_TESTING_MODE)
 
    def forced_recalibration(self, target_ppm: int) -> Optional[int]:
        """Perform forced recalibration"""
        if  target_ppm > 32000:
            return None
        if not self._write_data(self.STCC4_FORC_CALIBRATION, [target_ppm]):
            return None
            
        time.sleep(0.2)  # 200ms delay
        
        raw_data = self._read_data(self.STCC4_FORC_CALIBRATION, 3)
        if raw_data is None or len(raw_data) < 3:
            return None
            
        frc_correction = (raw_data[0] << 8) | raw_data[1]
        return frc_correction